polar-core/src/vm.rs

```diff
@@ -1517,15 +1463,16 @@ impl PolarVirtualMachine {
                 let left = args.pop().unwrap();
                 match (left.value(), right.value()) {
                     (Value::Variable(var), _) => match self.variable_state(var) {
-                        VariableState::Unbound | VariableState::Cycle(_) => {
-                            self.push_goal(Goal::Unify { left, right })?;
-                        }
                         VariableState::Bound(value) => {
                             return Err(self.type_error(&left, format!("Can only assign to unbound variables, {} is bound to value {}.", var.to_polar(), value.to_polar())));
                         }
+                        // TODO should this one be allowed?
                         VariableState::Partial(e) => {
                             return Err(self.type_error(&left, format!("Can only assign to unbound variables, {} is bound to expression {}.", var.to_polar(), e.to_polar())));
                         }
+                        _ => {
```

Comment:
Nit: Personally I think it's clearer to have the extra variant case spelled out.